[BE] typing for decorators

aorenste opened the pull request:
Summary:
Untyped decorators strip annotations from the decorated items.

- _compile
- _inductor/fx_passes/post_grad
- _inductor/lowering
- _library/custom_ops
- _meta_registrations
- _ops
- _refs/nn/functional
- ao/quantization/quantizer/xnnpack_quantizer_utils
- distributed/_composable/contract
- fx/experimental/graph_gradual_typechecker
- fx/experimental/migrate_gradual_types/constraint_generator
- optim/optimizer
- signal/windows/windows
- testing/_internal/common_device_type
- torch/_inductor/decomposition
- utils/flop_counter

Test Plan: unit tests

Differential Revision: D62302684


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

aorenste commented on torch/distributed/_composable/contract.py:
```diff
@@ -68,11 +123,19 @@ def contract(state_cls: Type[_State] = _State):
-    @wraps(state_cls)
-    def inner(func):
-        @wraps(func)
+    # @wraps(state_cls)
```
Oops - I think that was debugging cruft.

Skylion007 commented:
Partially addresses #142309

Skylion007 commented on torch/optim/optimizer.py:
```diff
@@ -113,15 +116,17 @@ def _stack_if_compiling(x):
 
-def _disable_dynamo_if_unsupported(single_tensor_fn=None):
+def _disable_dynamo_if_unsupported(
+    single_tensor_fn: Optional[Callable[..., object]] = None
```
```suggestion
    single_tensor_fn: Optional[Callable[_P], _T]] = None
```
or is there some reason that doesn't work?

aorenste replied:
It's not correct, is it? The single_tensor_fn isn't actually related to the returned value in any way (at least not directly in a way the decorator can see). For example look at the use of this on `adamax`. The `single_tensor_fn` has less varargs and more kwargs than the decorated function.

Skylion007 replied:
Ah good catch, could be its own paramsspec object though if the args/kwargs are useful

Skylion007 commented on torch/_ops.py:
```diff
@@ -15,6 +16,9 @@
+_P = ParamSpec("_P")
+
 
 _F = TypeVar("_F", bound=Callable[..., Any])
```
Get rid of `F_`, as it can still cause problems. Another PR is open to improve the situation

aorenste replied:
It seems like #144047 fixes this (existing) issue - so I'll land this and let that PR fix it.

aorenste committed:
[BE] typing for decorators (#144161)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/144161

Untyped decorators strip annotations from the decorated items.

- _compile
- _inductor/fx_passes/post_grad
- _inductor/lowering
- _library/custom_ops
- _meta_registrations
- _ops
- _refs/nn/functional
- ao/quantization/quantizer/xnnpack_quantizer_utils
- distributed/_composable/contract
- fx/experimental/graph_gradual_typechecker
- fx/experimental/migrate_gradual_types/constraint_generator
- optim/optimizer
- signal/windows/windows
- testing/_internal/common_device_type
- torch/_inductor/decomposition
- utils/flop_counter

Test Plan: unit tests

Differential Revision: D62302684

albanD reviewed (approved):
The nit from @Skylion007 needs fixing, sounds good once done.

aorenste commented:
@pytorchbot merge
